File: plugins/plugin-browser/typescript/src/actions/extract.ts
import type {
  Action,
  ActionResult,
  Content,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { ServiceType } from "@elizaos/core";
import type { BrowserService } from "../services/browser-service.js";
import {
  ActionError,
  handleBrowserError,
  ServiceNotAvailableError,
  SessionError,
} from "../utils/errors.js";

export const browserExtractAction: Action = {
  name: "BROWSER_EXTRACT",
  similes: ["EXTRACT_DATA", "GET_TEXT", "SCRAPE"],
  description: "Extract data from the webpage",

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const browserEnabled =
      runtime.getSetting("ENABLE_BROWSER") === "true" ||
      runtime.getSetting("BROWSER_ENABLED") === "true";

    if (!browserEnabled) {
      return false;
    }

    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      return false;
    }

    const text = message.content?.text?.toLowerCase() ?? "";
    return (
      text.includes("extract") ||
      text.includes("get") ||
      text.includes("scrape") ||
      text.includes("find") ||
      text.includes("read")
    );
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult | undefined> => {
    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      const error = new ServiceNotAvailableError();
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "extract data"
      );
      return {
        text: "Browser service is not available",
        success: false,
        data: {
          actionName: "BROWSER_EXTRACT",
          error: "service_not_available",
        },
        values: {
          success: false,
          errorType: "service_not_available",
        },
      };
    }

    const session = await service.getOrCreateSession();
    if (!session) {
      const error = new SessionError("No active browser session");
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "extract data"
      );
      return {
        text: "No active browser session",
        success: false,
        data: {
          actionName: "BROWSER_EXTRACT",
          error: "no_session",
        },
        values: {
          success: false,
          errorType: "no_session",
        },
      };
    }

    const text = message.content?.text ?? "";
    const match = text.match(/(?:extract|get|find|scrape|read) (?:the )?(.+?)(?:\s+from|\s*$)/i);
    const instruction = match ? match[1] : text;

    const result = await service.getClient().extract(session.id, instruction);
    if (!result.success) {
      throw new ActionError("extract", "page", new Error(result.error ?? "Extraction failed"));
    }

    const extractedData = result.data as { data?: string; found?: boolean } | undefined;
    const foundText = extractedData?.data ?? "No data found";
    const found = extractedData?.found ?? false;

    const responseContent: Content = {
      text: found
        ? `I found the ${instruction}: "${foundText}"`
        : `I couldn't find the requested ${instruction} on the page.`,
      actions: ["BROWSER_EXTRACT"],
      source: message.content?.source ?? "action",
    };

    await callback?.(responseContent);

    return {
      text: responseContent.text ?? "",
      success: true,
      data: {
        actionName: "BROWSER_EXTRACT",
        instruction,
        found,
        extractedData: foundText,
        sessionId: session.id,
      },
      values: {
        success: true,
        found,
        data: foundText,
      },
    };
  },

  examples: [
    [
      {
        name: "{{user}}",
        content: { text: "Extract the main heading from the page" },
      },
      {
        name: "{{agent}}",
        content: {
          text: 'I extracted the main heading: "Welcome to Our Website"',
          actions: ["BROWSER_EXTRACT"],
        },
      },
    ],
  ],
};

File: plugins/plugin-browser/rust/src/actions/extract.rs
use crate::services::BrowserService;
use crate::types::ActionResult;
use crate::utils::{action_error, parse_extract_instruction, session_error};
use std::collections::HashMap;
use std::sync::Arc;
use tracing::error;

pub const EXTRACT_ACTION_NAME: &str = "BROWSER_EXTRACT";
pub const EXTRACT_SIMILES: &[&str] = &["EXTRACT_DATA", "GET_TEXT", "SCRAPE"];
pub const EXTRACT_DESCRIPTION: &str = "Extract data from the webpage";

pub async fn browser_extract(service: Arc<BrowserService>, message: &str) -> ActionResult {
    let session = match service.get_or_create_session().await {
        Ok(s) => s,
        Err(e) => {
            let err = session_error(e);
            return ActionResult::failure(err.user_message);
        }
    };

    let instruction = parse_extract_instruction(message);

    let client = service.get_client();
    let result = client.extract(&session.id, &instruction).await;

    match result {
        Ok(response) if response.success => {
            let resp_data = response.data.unwrap_or_default();
            let found = resp_data
                .get("found")
                .and_then(|v| v.as_bool())
                .unwrap_or(false);
            let data_value = resp_data
                .get("data")
                .and_then(|v| v.as_str())
                .unwrap_or("No data found")
                .to_string();

            let mut data = HashMap::new();
            data.insert(
                "actionName".to_string(),
                serde_json::json!(EXTRACT_ACTION_NAME),
            );
            data.insert("instruction".to_string(), serde_json::json!(instruction));
            data.insert("found".to_string(), serde_json::json!(found));
            data.insert("data".to_string(), serde_json::json!(data_value));
            data.insert("sessionId".to_string(), serde_json::json!(session.id));

            ActionResult::success(data)
        }
        Ok(response) => {
            let err = action_error("extract", "page", response.error.as_deref());
            ActionResult::failure(err.user_message)
        }
        Err(e) => {
            error!("Error in BROWSER_EXTRACT action: {}", e);
            let err = action_error("extract", "page", Some(&e));
            ActionResult::failure(err.user_message)
        }
    }
}
